File: ScaleformUI_Csharp/Menus/Pause Menus/Elements/Columns/SubmenuCentralColumn.cs
using ScaleformUI.Menus;
using ScaleformUI.PauseMenu;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ScaleformUI.PauseMenus.Elements
{
    public class SubmenuCentralColumn : PM_Column
    {
        internal LeftItemType currentColumnType = LeftItemType.Empty;

        public SubmenuCentralColumn(int position) : base(position)
        {
        }

        public override void SetDataSlot(int index)
        {
            TabLeftItem curItem = (TabLeftItem)Parent.LeftColumn.Items[Parent.LeftColumn.Index];
            switch (curItem.ItemType)
            {
                case LeftItemType.Info:
                    {
                        var _item = Items[index];
                        var labels = _item.Label.SplitLabel;
                        BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "SET_DATA_SLOT");
                        PushScaleformMovieFunctionParameterInt((int)position);
                        PushScaleformMovieFunctionParameterInt(index);
                        PushScaleformMovieFunctionParameterInt(0);
                        PushScaleformMovieFunctionParameterInt(index);
                        PushScaleformMovieFunctionParameterInt(0);
                        PushScaleformMovieFunctionParameterInt(0);
                        PushScaleformMovieFunctionParameterBool(true);
                        BeginTextCommandScaleformString("CELL_EMAIL_BCON");
                        for (var i = 0; i < labels?.Length; i++)
                            AddTextComponentScaleform(labels[i]);
                        EndTextCommandScaleformString_2();
                        EndScaleformMovieMethod();
                    }
                    break;
                case LeftItemType.Statistics:
                    {
                        StatsTabItem it = (StatsTabItem)Items[index];
                        BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "SET_DATA_SLOT");
                        PushScaleformMovieFunctionParameterInt((int)position);
                        PushScaleformMovieFunctionParameterInt(index);
                        PushScaleformMovieFunctionParameterInt(0);
                        PushScaleformMovieFunctionParameterInt(index);
                        PushScaleformMovieFunctionParameterInt((int)it.Type);
                        PushScaleformMovieFunctionParameterInt(0);
                        PushScaleformMovieFunctionParameterBool(true);
                        PushScaleformMovieFunctionParameterString(it.Label.Label);
                        switch (it.Type)
                        {
                            case StatItemType.Basic:
                                PushScaleformMovieFunctionParameterString(it.RightLabel);
                                break;
                            case StatItemType.ColoredBar:
                                PushScaleformMovieMethodParameterInt(it.Value);
                                PushScaleformMovieMethodParameterInt(it.ColoredBarColor.ArgbValue);
                                break;
                        }
                        EndScaleformMovieMethod();
                    }
                    break;
                case LeftItemType.Settings:
                    {
                        SettingsItem it = (SettingsItem)Items[index];
                        BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "SET_DATA_SLOT");
                        PushScaleformMovieFunctionParameterInt((int)position);
                        PushScaleformMovieFunctionParameterInt(index);
                        PushScaleformMovieFunctionParameterInt(0);
                        PushScaleformMovieFunctionParameterInt(index);
                        PushScaleformMovieFunctionParameterInt((int)it.ItemType);
                        switch (it.ItemType)
                        {
                            case SettingsItemType.ListItem:
                                PushScaleformMovieFunctionParameterInt(((SettingsListItem)it).ItemIndex);
                                break;
                            case SettingsItemType.SliderBar:
                                PushScaleformMovieFunctionParameterInt(((SettingsSliderItem)it).Value);
                                break;
                            case SettingsItemType.MaskedProgressBar:
                            case SettingsItemType.ProgressBar:
                                PushScaleformMovieFunctionParameterInt(((SettingsProgressItem)it).Value);
                                break;
                            default:
                                PushScaleformMovieFunctionParameterInt(0);
                                break;
                        }
                        PushScaleformMovieFunctionParameterBool(true);
                        if (it.ItemType == SettingsItemType.BlipType)
                        {
                            BeginTextCommandScaleformString("STRING");
                            AddTextComponentScaleform(it.Label.Label);
                            EndTextCommandScaleformString_2();
                        }
                        else
                        {
                            PushScaleformMovieFunctionParameterString(it.Label.Label);
                        }
                        switch (it.ItemType)
                        {
                            case SettingsItemType.Basic:
                                PushScaleformMovieFunctionParameterString(it.RightLabel);
                                break;
                            case SettingsItemType.ListItem:
                                PushScaleformMovieFunctionParameterString(string.Join(",", ((SettingsListItem)it).ListItems));
                                break;
                            case SettingsItemType.CheckBox:
                                PushScaleformMovieFunctionParameterInt((int)((SettingsCheckboxItem)it).CheckBoxStyle);
                                PushScaleformMovieFunctionParameterBool(((SettingsCheckboxItem)it).IsChecked);
                                break;
                            case SettingsItemType.MaskedProgressBar:
                            case SettingsItemType.ProgressBar:
                                PushScaleformMovieMethodParameterInt(((SettingsProgressItem)it).MaxValue);
                                PushScaleformMovieMethodParameterInt(((SettingsProgressItem)it).ColoredBarColor.ArgbValue);
                                break;
                            case SettingsItemType.SliderBar:
                                PushScaleformMovieMethodParameterInt(((SettingsSliderItem)it).MaxValue);
                                PushScaleformMovieMethodParameterInt(((SettingsSliderItem)it).ColoredBarColor.ArgbValue);
                                break;
                            case SettingsItemType.BlipType:
                                PushScaleformMovieFunctionParameterString(it.RightLabel);
                                break;
                        }
                        EndScaleformMovieMethod();
                        SetColumnScroll(Index + 1, Items.Count, VisibleItems, string.Empty, Items.Count < VisibleItems);
                    }
                    break;
                case LeftItemType.Keymap:
                    {
                        var _item = (KeymapItem)Items[index];
                        BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "SET_DATA_SLOT");
                        PushScaleformMovieFunctionParameterInt((int)position);
                        PushScaleformMovieFunctionParameterInt(index);
                        PushScaleformMovieFunctionParameterInt(0);
                        PushScaleformMovieFunctionParameterInt(index);
                        PushScaleformMovieFunctionParameterInt(0);
                        PushScaleformMovieFunctionParameterInt(0);
                        PushScaleformMovieFunctionParameterBool(true);
                        BeginTextCommandScaleformString("STRING");
                        AddTextComponentScaleform(_item.Label.Label);
                        EndTextCommandScaleformString_2();
                        BeginTextCommandScaleformString("STRING");
                        AddTextComponentScaleform(Game.CurrentInputMode == InputMode.MouseAndKeyboard ? _item.PrimaryKeyboard : _item.PrimaryGamepad);
                        EndTextCommandScaleformString_2();
                        BeginTextCommandScaleformString("STRING");
                        AddTextComponentScaleform(Game.CurrentInputMode == InputMode.MouseAndKeyboard ? _item.SecondaryKeyboard : _item.SecondaryGamepad);
                        EndTextCommandScaleformString_2();
                        EndScaleformMovieMethod();
                    }
                    break;
            }
        }

        public override void UpdateSlot(int index)
        {
            switch (currentColumnType)
            {
                case LeftItemType.Info:
                    {
                        var _item = Items[index];
                        var labels = _item.Label.SplitLabel;
                        BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "UPDATE_SLOT");
                        PushScaleformMovieFunctionParameterInt((int)position);
                        PushScaleformMovieFunctionParameterInt(index);
                        PushScaleformMovieFunctionParameterInt(0);
                        PushScaleformMovieFunctionParameterInt(index);
                        PushScaleformMovieFunctionParameterInt(0);
                        PushScaleformMovieFunctionParameterInt(0);
                        PushScaleformMovieFunctionParameterBool(true);
                        BeginTextCommandScaleformString("CELL_EMAIL_BCON");
                        for (var i = 0; i < labels?.Length; i++)
                            AddTextComponentScaleform(labels[i]);
                        EndTextCommandScaleformString_2();
                        EndScaleformMovieMethod();
                    }
                    break;
                case LeftItemType.Statistics:
                    {
                        StatsTabItem it = (StatsTabItem)Items[index];
                        BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "UPDATE_SLOT");
                        PushScaleformMovieFunctionParameterInt((int)position);
                        PushScaleformMovieFunctionParameterInt(index);
                        PushScaleformMovieFunctionParameterInt(0);
                        PushScaleformMovieFunctionParameterInt(index);
                        PushScaleformMovieFunctionParameterInt((int)it.Type);
                        PushScaleformMovieFunctionParameterInt(0);
                        PushScaleformMovieFunctionParameterBool(true);
                        PushScaleformMovieFunctionParameterString(it.Label.Label);
                        switch (it.Type)
                        {
                            case StatItemType.Basic:
                                PushScaleformMovieFunctionParameterString(it.RightLabel);
                                break;
                            case StatItemType.ColoredBar:
                                PushScaleformMovieMethodParameterInt(it.Value);
                                PushScaleformMovieMethodParameterInt(it.ColoredBarColor.ArgbValue);
                                break;
                        }
                        EndScaleformMovieMethod();
                    }
                    break;
                case LeftItemType.Settings:
                    {
                        SettingsItem it = (SettingsItem)Items[index];
                        BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "UPDATE_SLOT");
                        PushScaleformMovieFunctionParameterInt((int)position);
                        PushScaleformMovieFunctionParameterInt(index);
                        PushScaleformMovieFunctionParameterInt(0);
                        PushScaleformMovieFunctionParameterInt(index);
                        PushScaleformMovieFunctionParameterInt((int)it.ItemType);
                        switch (it.ItemType)
                        {
                            case SettingsItemType.ListItem:
                                PushScaleformMovieFunctionParameterInt(((SettingsListItem)it).ItemIndex);
                                break;
                            case SettingsItemType.SliderBar:
                                PushScaleformMovieFunctionParameterInt(((SettingsSliderItem)it).Value);
                                break;
                            case SettingsItemType.MaskedProgressBar:
                            case SettingsItemType.ProgressBar:
                                PushScaleformMovieFunctionParameterInt(((SettingsProgressItem)it).Value);
                                break;
                            default:
                                PushScaleformMovieFunctionParameterInt(0);
                                break;
                        }
                        PushScaleformMovieFunctionParameterBool(true);
                        if (it.ItemType == SettingsItemType.BlipType)
                        {
                            BeginTextCommandScaleformString("STRING");
                            AddTextComponentScaleform(it.Label.Label);
                            EndTextCommandScaleformString_2();
                        }
                        else
                        {
                            PushScaleformMovieFunctionParameterString(it.Label.Label);
                        }
                        switch (it.ItemType)
                        {
                            case SettingsItemType.Basic:
                                PushScaleformMovieFunctionParameterString(it.RightLabel);
                                break;
                            case SettingsItemType.ListItem:
                                PushScaleformMovieFunctionParameterString(string.Join(",", ((SettingsListItem)it).ListItems));
                                break;
                            case SettingsItemType.CheckBox:
                                PushScaleformMovieFunctionParameterInt((int)((SettingsCheckboxItem)it).CheckBoxStyle);
                                PushScaleformMovieFunctionParameterBool(((SettingsCheckboxItem)it).IsChecked);
                                break;
                            case SettingsItemType.MaskedProgressBar:
                            case SettingsItemType.ProgressBar:
                                PushScaleformMovieMethodParameterInt(((SettingsProgressItem)it).MaxValue);
                                PushScaleformMovieMethodParameterInt(((SettingsProgressItem)it).ColoredBarColor.ArgbValue);
                                break;
                            case SettingsItemType.SliderBar:
                                PushScaleformMovieMethodParameterInt(((SettingsSliderItem)it).MaxValue);
                                PushScaleformMovieMethodParameterInt(((SettingsSliderItem)it).ColoredBarColor.ArgbValue);
                                break;
                            case SettingsItemType.BlipType:
                                PushScaleformMovieFunctionParameterString(it.RightLabel);
                                break;
                        }
                        EndScaleformMovieMethod();
                        SetColumnScroll(Index + 1, Items.Count, VisibleItems, string.Empty, Items.Count < VisibleItems);
                    }
                    break;
                case LeftItemType.Keymap:
                    {
                        var _item = (KeymapItem)Items[index];
                        BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "UPDATE_SLOT");
                        PushScaleformMovieFunctionParameterInt((int)position);
                        PushScaleformMovieFunctionParameterInt(index);
                        PushScaleformMovieFunctionParameterInt(0);
                        PushScaleformMovieFunctionParameterInt(index);
                        PushScaleformMovieFunctionParameterInt(0);
                        PushScaleformMovieFunctionParameterInt(0);
                        PushScaleformMovieFunctionParameterBool(true);
                        BeginTextCommandScaleformString("STRING");
                        AddTextComponentScaleform(_item.Label.Label);
                        EndTextCommandScaleformString_2();
                        BeginTextCommandScaleformString("STRING");
                        AddTextComponentScaleform(Game.CurrentInputMode == InputMode.MouseAndKeyboard ? _item.PrimaryKeyboard : _item.PrimaryGamepad);
                        EndTextCommandScaleformString_2();
                        BeginTextCommandScaleformString("STRING");
                        AddTextComponentScaleform(Game.CurrentInputMode == InputMode.MouseAndKeyboard ? _item.SecondaryKeyboard : _item.SecondaryGamepad);
                        EndTextCommandScaleformString_2();
                        EndScaleformMovieMethod();
                    }
                    break;
            }
        }

        public virtual void AddSlot(int index) { }

        public override async void GoUp()
        {
            if (currentColumnType == LeftItemType.Settings)
            {
                ((SettingsItem)Items[Index]).Selected = false;
                do
                {
                    index--;
                    if(index < 0)
                        index = Items.Count - 1;
                    await BaseScript.Delay(0);
                } while (((SettingsItem)Items[Index]).ItemType == SettingsItemType.Empty || ((SettingsItem)Items[Index]).ItemType == SettingsItemType.Separator);
                ((SettingsItem)Items[Index]).Selected = true;
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_HIGHLIGHT", (int)position, index, true, true);
                SetColumnScroll(Index + 1, Items.Count, VisibleItems, string.Empty, Items.Count < VisibleItems);
            }
        }

        public override async void GoDown()
        {
            if (currentColumnType == LeftItemType.Settings)
            {
                ((SettingsItem)Items[Index]).Selected = false;
                do
                {
                    index++;
                    if (index >= Items.Count)
                        index = 0;
                    await BaseScript.Delay(0);
                } while (((SettingsItem)Items[Index]).ItemType == SettingsItemType.Empty || ((SettingsItem)Items[Index]).ItemType == SettingsItemType.Separator);
                ((SettingsItem)Items[Index]).Selected = true;
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_HIGHLIGHT", (int)position, index, true, true);
                SetColumnScroll(Index + 1, Items.Count, VisibleItems, string.Empty, Items.Count < VisibleItems);
            }
        }

        public virtual void GoLeft()
        {
            if (currentColumnType == LeftItemType.Settings)
            {
                var item = (SettingsItem)Items[Index];
                switch (item.ItemType)
                {
                    case SettingsItemType.ListItem:
                        (item as SettingsListItem).itemIndex--;
                        (item as SettingsListItem).ListChanged();
                        break;
                    case SettingsItemType.SliderBar: //TODO: UPDATE WITH MULTIPLIER
                        (item as SettingsSliderItem)._value--;
                        (item as SettingsSliderItem).SliderChanged();
                        break;
                    case SettingsItemType.ProgressBar:
                    case SettingsItemType.MaskedProgressBar: //TODO: UPDATE WITH MULTIPLIER
                        (item as SettingsProgressItem)._value--;
                        (item as SettingsProgressItem).ProgressChanged();
                        break;
                }
            }
        }

        public virtual void GoRight() 
        {
            if(currentColumnType == LeftItemType.Settings)
            {
                var item = (SettingsItem)Items[Index];
                switch (item.ItemType)
                {
                    case SettingsItemType.ListItem:
                        (item as SettingsListItem).itemIndex++;
                        (item as SettingsListItem).ListChanged();
                        break;
                    case SettingsItemType.SliderBar: //TODO: UPDATE WITH MULTIPLIER
                        (item as SettingsSliderItem)._value++;
                        (item as SettingsSliderItem).SliderChanged();
                        break;
                    case SettingsItemType.ProgressBar:
                    case SettingsItemType.MaskedProgressBar: //TODO: UPDATE WITH MULTIPLIER
                        (item as SettingsProgressItem)._value++;
                        (item as SettingsProgressItem).ProgressChanged();
                        break;
                }
            }
        }

        public virtual void Select() {
            if (Items[Index] is SettingsItem item)
            {
                if (!item.Enabled)
                {
                    Game.PlaySound(TabView.AUDIO_ERROR, TabView.AUDIO_LIBRARY);
                    return;
                }
                switch (item.ItemType)
                {
                    case SettingsItemType.ListItem:
                        (item as SettingsListItem).ListSelected();
                        break;
                    case SettingsItemType.CheckBox:
                        (item as SettingsCheckboxItem).IsChecked = !(item as SettingsCheckboxItem).IsChecked;
                        break;
                    case SettingsItemType.MaskedProgressBar:
                    case SettingsItemType.ProgressBar:
                        (item as SettingsProgressItem).ProgressSelected();
                        break;
                    case SettingsItemType.SliderBar:
                        (item as SettingsSliderItem).SliderSelected();
                        break;
                }
                // we always trigger this because all items inherit this except checkbox
                if (item.ItemType != SettingsItemType.CheckBox)
                    item.Activated();
            }

        }

        public override void MouseScroll(int dir)
        {
            //Parent._pause._pause.CallFunction("DELTA_MOUSE_WHEEL", dir);
        }
    }
}

File: ScaleformUI_Lua/src/Menus/PauseMenus/Elements/Columns/SubmenuCentralColumn.lua
SubmenuCentralColumn = {}
SubmenuCentralColumn.__index = SubmenuCentralColumn
setmetatable(SubmenuCentralColumn, { __index = PM_Column })
SubmenuCentralColumn.__call = function() return "SubmenuCentralColumn" end

function SubmenuCentralColumn.New(pos)
    local col = PM_Column.New(pos)
    col._type = 0
    return setmetatable(col, SubmenuCentralColumn)
end

function SubmenuCentralColumn:currentColumnType(t)
    if t == nil then
        return self._type
    else
        self._type = t
    end
end

function SubmenuCentralColumn:SetDataSlot(index)
    local curItem = self.Parent.LeftColumn.Items[self.Parent.LeftColumn:Index()]
    local _item = self.Items[index]
    if curItem.ItemType == LeftItemType.Info then
        local label = "SUBMN_CCOL_LBL"
        BeginScaleformMovieMethod(ScaleformUI.Scaleforms._pauseMenu._pause.handle, "SET_DATA_SLOT")
        PushScaleformMovieFunctionParameterInt(self.position)
        PushScaleformMovieFunctionParameterInt(index - 1)
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterInt(index - 1)
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterBool(true)
        AddTextEntry(label, _item.label)
        BeginTextCommandScaleformString(label)
        EndTextCommandScaleformString_2()
        EndScaleformMovieMethod()
    elseif curItem.ItemType == LeftItemType.Statistics then
        BeginScaleformMovieMethod(ScaleformUI.Scaleforms._pauseMenu._pause.handle, "SET_DATA_SLOT")
        PushScaleformMovieFunctionParameterInt(self.position)
        PushScaleformMovieFunctionParameterInt(index - 1)
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterInt(index - 1)
        PushScaleformMovieFunctionParameterInt(_item.Type)
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterBool(true)
        PushScaleformMovieFunctionParameterString(_item.label)
        if _item.Type == StatItemType.Basic then
            PushScaleformMovieFunctionParameterString(_item._rightLabel)
        elseif _item.Type == StatItemType.ColoredBar then
            PushScaleformMovieMethodParameterInt(_item._value)
            PushScaleformMovieMethodParameterInt(_item._coloredBarColor:ToArgb())
        end
        EndScaleformMovieMethod()
    elseif curItem.ItemType == LeftItemType.Settings then
        BeginScaleformMovieMethod(ScaleformUI.Scaleforms._pauseMenu._pause.handle, "SET_DATA_SLOT")
        PushScaleformMovieFunctionParameterInt(self.position)
        PushScaleformMovieFunctionParameterInt(index - 1)
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterInt(index - 1)
        PushScaleformMovieFunctionParameterInt(_item.ItemType)
        if _item.ItemType == SettingsItemType.ListItem then
            PushScaleformMovieFunctionParameterInt(_item._itemIndex)
        elseif _item.ItemType == SettingsItemType.SliderBar or _item.ItemType == SettingsItemType.MaskedProgressBar or _item.ItemType == SettingsItemType.ProgressBar then
            PushScaleformMovieFunctionParameterInt(_item._value)
        else
            PushScaleformMovieFunctionParameterInt(0)
        end
        PushScaleformMovieFunctionParameterBool(true)
        if (_item.ItemType == SettingsItemType.BlipType) then
            BeginTextCommandScaleformString("STRING")
            AddTextComponentScaleform(_item.label)
            EndTextCommandScaleformString_2()
        else
            PushScaleformMovieFunctionParameterString(_item.label)
        end
        if _item.ItemType == SettingsItemType.Basic then
            PushScaleformMovieFunctionParameterString(_item._rightLabel)
        elseif _item.ItemType == SettingsItemType.ListItem then
            PushScaleformMovieFunctionParameterString(Join(",", _item.ListItems))
        elseif _item.ItemType == SettingsItemType.CheckBox then
            PushScaleformMovieFunctionParameterInt(_item.CheckBoxStyle)
            PushScaleformMovieFunctionParameterBool(_item._isChecked)
        elseif _item.ItemType == SettingsItemType.MaskedProgressBar or _item.ItemType == SettingsItemType.ProgressBar or _item.ItemType == SettingsItemType.SliderBar then
            PushScaleformMovieMethodParameterInt(_item.MaxValue)
            PushScaleformMovieMethodParameterInt(_item._coloredBarColor:ToArgb())
        elseif _item.ItemType == SettingsItemType.BlipType then
            PushScaleformMovieFunctionParameterString(_item._rightLabel)
        end
        EndScaleformMovieMethod()
        self:SetColumnScroll(self.index, #self.Items, self.VisibleItems, "", #self.Items < self.VisibleItems)
    elseif curItem.ItemType == LeftItemType.Keymap then
        BeginScaleformMovieMethod(ScaleformUI.Scaleforms._pauseMenu._pause.handle, "SET_DATA_SLOT")
        PushScaleformMovieFunctionParameterInt(self.position)
        PushScaleformMovieFunctionParameterInt(index - 1)
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterInt(index - 1)
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterBool(true)
        BeginTextCommandScaleformString("STRING")
        AddTextComponentScaleform(_item.label)
        EndTextCommandScaleformString_2()
        BeginTextCommandScaleformString("STRING")
        if IsUsingKeyboard(2) then
            AddTextComponentScaleform(_item.PrimaryKeyboard)
        else
            AddTextComponentScaleform(_item.PrimaryGamepad)
        end
        EndTextCommandScaleformString_2()
        BeginTextCommandScaleformString("STRING")
        if IsUsingKeyboard(2) then
            AddTextComponentScaleform(_item.SecondaryKeyboard)
        else
            AddTextComponentScaleform(_item.SecondaryGamepad)
        end
        EndTextCommandScaleformString_2()
        EndScaleformMovieMethod()
    end
end

function SubmenuCentralColumn:UpdateSlot(index)
    local _item = self.Items[index]
    if self:currentColumnType() == LeftItemType.Info then
        local label = "SUBMN_CCOL_LBL"
        BeginScaleformMovieMethod(ScaleformUI.Scaleforms._pauseMenu._pause.handle, "UPDATE_SLOT")
        PushScaleformMovieFunctionParameterInt(self.position)
        PushScaleformMovieFunctionParameterInt(index - 1)
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterInt(index - 1)
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterBool(true)
        AddTextEntry(label, _item.label)
        BeginTextCommandScaleformString(label)
        EndTextCommandScaleformString_2()
        EndScaleformMovieMethod()
    elseif self:currentColumnType() == LeftItemType.Statistics then
        BeginScaleformMovieMethod(ScaleformUI.Scaleforms._pauseMenu._pause.handle, "UPDATE_SLOT")
        PushScaleformMovieFunctionParameterInt(self.position)
        PushScaleformMovieFunctionParameterInt(index - 1)
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterInt(index - 1)
        PushScaleformMovieFunctionParameterInt(_item.Type)
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterBool(true)
        PushScaleformMovieFunctionParameterString(_item.label)
        if _item.Type == StatItemType.Basic then
            PushScaleformMovieFunctionParameterString(_item._rightLabel)
        elseif _item.Type == StatItemType.ColoredBar then
            PushScaleformMovieMethodParameterInt(_item._value)
            PushScaleformMovieMethodParameterInt(_item._coloredBarColor:ToArgb())
        end
        EndScaleformMovieMethod()
    elseif self:currentColumnType() == LeftItemType.Settings then
        BeginScaleformMovieMethod(ScaleformUI.Scaleforms._pauseMenu._pause.handle, "UPDATE_SLOT")
        PushScaleformMovieFunctionParameterInt(self.position)
        PushScaleformMovieFunctionParameterInt(index - 1)
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterInt(index - 1)
        PushScaleformMovieFunctionParameterInt(_item.ItemType)
        if _item.ItemType == SettingsItemType.ListItem then
            PushScaleformMovieFunctionParameterInt(_item.ItemIndex)
        elseif _item.ItemType == SettingsItemType.SliderBar or _item.ItemType == SettingsItemType.MaskedProgressBar or _item.ItemType == SettingsItemType.ProgressBar then
            PushScaleformMovieFunctionParameterInt(_item.Value)
        else
            PushScaleformMovieFunctionParameterInt(0)
        end
        PushScaleformMovieFunctionParameterBool(true)
        if (_item.ItemType == SettingsItemType.BlipType) then
            BeginTextCommandScaleformString("STRING")
            AddTextComponentScaleform(_item.label)
            EndTextCommandScaleformString_2()
        else
            PushScaleformMovieFunctionParameterString(_item.label)
        end
        if _item.ItemType == SettingsItemType.Basic then
            PushScaleformMovieFunctionParameterString(_item.RightLabel)
        elseif _item.ItemType == SettingsItemType.ListItem then
            PushScaleformMovieFunctionParameterString(Join(",", _item.ListItems))
        elseif _item.ItemType == SettingsItemType.CheckBox then
            PushScaleformMovieFunctionParameterInt(_item.CheckBoxStyle)
            PushScaleformMovieFunctionParameterBool(_item.IsChecked)
        elseif _item.ItemType == SettingsItemType.MaskedProgressBar or _item.ItemType == SettingsItemType.ProgressBar or  _item.ItemType == SettingsItemType.SliderBar then
            PushScaleformMovieMethodParameterInt(_item.MaxValue)
            PushScaleformMovieMethodParameterInt(_item.ColoredBarColor:ToArgb())
        elseif _item.ItemType == SettingsItemType.BlipType then
            PushScaleformMovieFunctionParameterString(_item.RightLabel)
        end
        EndScaleformMovieMethod()
        self:SetColumnScroll(self.index, #self.Items, self.VisibleItems, "", #self.Items < self.VisibleItems)
    elseif self:currentColumnType() == LeftItemType.Keymap then
        BeginScaleformMovieMethod(ScaleformUI.Scaleforms._pauseMenu._pause.handle, "UPDATE_SLOT")
        PushScaleformMovieFunctionParameterInt(self.position)
        PushScaleformMovieFunctionParameterInt(index - 1)
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterInt(index - 1)
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterBool(true)
        BeginTextCommandScaleformString("STRING")
        AddTextComponentScaleform(_item.label)
        EndTextCommandScaleformString_2()
        BeginTextCommandScaleformString("STRING")
        if IsUsingKeyboard(2) then
            AddTextComponentScaleform(_item.PrimaryKeyboard)
        else
            AddTextComponentScaleform(_item.PrimaryGamepad)
        end
        EndTextCommandScaleformString_2()
        BeginTextCommandScaleformString("STRING")
        if IsUsingKeyboard(2) then
            AddTextComponentScaleform(_item.SecondaryKeyboard)
        else
            AddTextComponentScaleform(_item.SecondaryGamepad)
        end
        EndTextCommandScaleformString_2()
        EndScaleformMovieMethod()
    end
end

function SubmenuCentralColumn:GoUp()
    if self:currentColumnType() == LeftItemType.Settings then
        Citizen.CreateThread(function()
            self.Items[self.index]:Selected(false)
            repeat
                self.index = self.index - 1
                if self.index < 1 then
                    self.index = #self.Items
                end
                Wait(0)
            until self.Items[self.index].ItemType ~= SettingsItemType.Separator and self.Items[self.index].ItemType ~= SettingsItemType.Empty
            self.Items[self.index]:Selected(true)
            ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_HIGHLIGHT", self.position, self.index - 1,
                true, true)
            self:SetColumnScroll(self.index, #self.Items, self.VisibleItems, "", #self.Items < self.VisibleItems)
        end)
    end
end

function SubmenuCentralColumn:GoDown()
    if self:currentColumnType() == LeftItemType.Settings then
        Citizen.CreateThread(function()
            self.Items[self.index]:Selected(false)
            repeat
                self.index = self.index + 1
                if self.index > #self.Items then
                    self.index = 1
                end
                Wait(0)
            until self.Items[self.index].ItemType ~= SettingsItemType.Separator and self.Items[self.index].ItemType ~= SettingsItemType.Empty
            self.Items[self.index]:Selected(true)
            ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_HIGHLIGHT", self.position, self.index - 1,
                true, true)
            self:SetColumnScroll(self.index, #self.Items, self.VisibleItems, "", #self.Items < self.VisibleItems)
        end)
    end
end

function SubmenuCentralColumn:GoLeft()
    if self:currentColumnType() == LeftItemType.Settings then
        local item = self.Items[self.index]
        if item.ItemType == SettingsItemType.ListItem then
            item._itemIndex = item._itemIndex - 1
            item.OnListChanged(item, item:ItemIndex(), tostring(item.ListItems[item:ItemIndex()]))
        elseif item.ItemType == SettingsItemType.SliderBar or item.ItemType == SettingsItemType.ProgressBar or item.ItemType == SettingsItemType.MaskedProgressBar then
            item._value = item._value - 1
            item.OnBarChanged(item, item._value)
        end
    end
end

function SubmenuCentralColumn:GoRight()
    if self:currentColumnType() == LeftItemType.Settings then
        local item = self.Items[self.index]
        if item.ItemType == SettingsItemType.ListItem then
            item._itemIndex = item._itemIndex + 1
            item.OnListChanged(item, item:ItemIndex(), tostring(item.ListItems[item:ItemIndex()]))
        elseif item.ItemType == SettingsItemType.SliderBar or item.ItemType == SettingsItemType.ProgressBar or item.ItemType == SettingsItemType.MaskedProgressBar then
            item._value = item._value + 1
            item.OnBarChanged(item, item._value)
        end
    end
end

function SubmenuCentralColumn:Select()
    if self:currentColumnType() == LeftItemType.Settings then
        local item = self.Items[self.index]
        if not item:Enabled() then
            PlaySoundFrontend(-1, "ERROR", "HUD_FRONTEND_DEFAULT_SOUNDSET", false)
            return
        end
        if item.ItemType == SettingsItemType.ListItem then
            item.OnListSelected(item, item:ItemIndex(), tostring(item.ListItems[item:ItemIndex()]))
        elseif item.ItemType == SettingsItemType.CheckBox then
            item:Checked(not item:Checked())
        elseif item.ItemType == SettingsItemType.SliderBar then
            item.OnSliderSelected(item, item._value)
        elseif item.ItemType == SettingsItemType.ProgressBar or item.ItemType == SettingsItemType.MaskedProgressBar then
            item.OnProgressSelected(item, item._value)
        end
        if item.ItemType ~= SettingsItemType.CheckBox then
            item.OnActivated()
        end
    end
end
